File: 0041-first-missing-positive/0041-first-missing-positive.go
func firstMissingPositive(nums []int) int {
	for i, _ := range nums {
		for nums[i] != 0 && nums[i]-1 != i {
			if nums[i]-1 < 0 || nums[i]-1 >= len(nums) || nums[nums[i]-1] == nums[i] {
				nums[i] = 0
			} else {
				nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
			}
		}
	}
	for i, num := range nums {
		if num == 0 {
			return i + 1
		}
	}
	return len(nums) + 1
}

File: 0041-first-missing-positive/0041-first-missing-positive.py
class Solution:
  def firstMissingPositive(self, nums: List[int]) -> int:
    n = len(nums)
    for i in range(n):
      while nums[i] and nums[i]-1 != i:
        if nums[i] <= 0 or nums[i] > n or nums[nums[i]-1] == nums[i]:
          nums[i] = 0
        else:
          nums[nums[i]-1], nums[i] = nums[i], nums[nums[i]-1]
    for i in range(n):
      if not nums[i]:
        return i + 1
    return n + 1
        